### PR TITLE
Add component metadata and subgraph evaluation.

### PR DESCRIPTION
**Pull request by csams**
Component types and declarations can be enriched with arbitrary
metadata by declaring a type like component = new_component_type(type_metadata=)
or directly in the component declaration with @component(metadata=...).
The metadata in the component declaration inherits from the type and overrides
keys in case of collision.

dr now includes get_subgraphs which takes a dependency graph and yields all
of the disjoint subgraphs within it. This functionality is used within
run_incremental to yield brokers for each independent subgraph given a starting
graph. This should help with memory consumption when executing large numbers of
mostly independent components.

### CODE REVIEW

File: insights/core/dr.py [hunk 1]
```diff
@@ -139,7 +163,7 @@ def stringify_requirements(requires):
 
 
 def register_component(component, delegate, component_type,
-        requires=None, optional=None, group=GROUPS.single, alias=None):
+        requires=None, optional=None, group=GROUPS.single, alias=None, metadata={}):
```
Comment (jhjaggars) [inline insights/core/dr.py]: the use of `{}` rather than `dict()` is inconsistent here.

File: insights/core/dr.py [hunk 2]
```diff
@@ -46,6 +50,26 @@ def get_name(component):
     return str(component)
 
 
+def get_metadata(component):
+    return COMPONENT_METADATA.get(component)
+
+
+def get_subgraphs(graph=DEPENDENCIES):
+    keys = set(graph.keys())
```
Comment (jhjaggars) [inline insights/core/dr.py]: I think `set(graph)` behaves the same and saves you a copy.

File: insights/core/dr.py [hunk 3]
```diff
@@ -46,6 +50,26 @@ def get_name(component):
...
+    seen = set()
+    while keys:
+        frontier = {keys.pop()}
```
Comment (jhjaggars) [inline insights/core/dr.py]: I think you'll have to use `set([keys.pop()])` to support python 2.6.

Comment (csams) [inline insights/core/dr.py]: Good catch.

**Pull request merged**